staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_filter_test.go

```diff
@@ -0,0 +1,166 @@
+				BorrowingLimitPercent:    &borrowingLimit,
+				LimitResponse: flowcontrol.LimitResponse{
+					Type: flowcontrol.LimitResponseTypeQueue,
+					Queuing: &flowcontrol.QueuingConfiguration{
+						Queues:           10,
+						HandSize:         2,
+						QueueLengthLimit: 10,
+					},
+				},
+			},
+		},
+	}
+	cfgObjs = append(cfgObjs, fsObj, plcObj)
+
+	clientset := clientsetfake.NewSimpleClientset(cfgObjs...)
+	informerFactory := informers.NewSharedInformerFactory(clientset, time.Second)
+	flowcontrolClient := clientset.FlowcontrolV1beta3()
+	clk := eventclock.Real{}
+	controller := newTestableController(TestableConfig{
+		Name:                   "Controller",
+		Clock:                  clk,
+		AsFieldManager:         ConfigConsumerAsFieldManager,
+		FoundToDangling:        func(found bool) bool { return !found },
+		InformerFactory:        informerFactory,
+		FlowcontrolClient:      flowcontrolClient,
+		ServerConcurrencyLimit: 24,
+		RequestWaitLimit:       time.Minute,
+		ReqsGaugeVec:           metrics.PriorityLevelConcurrencyGaugeVec,
+		ExecSeatsGaugeVec:      metrics.PriorityLevelExecutionSeatsGaugeVec,
+		QueueSetFactory:        fqs.NewQueueSetFactory(clk),
+	})
+
+	informerFactory.Start(nil)
+
+	status := informerFactory.WaitForCacheSync(nil)
+	if names := unsynced(status); len(names) > 0 {
+		t.Fatalf("WaitForCacheSync did not successfully complete, resources=%#v", names)
+	}
+
+	go func() {
+		controller.Run(nil)
+	}()
+
+	// ensure that the controller has run its first loop.
+	err := wait.PollImmediate(100*time.Millisecond, 5*time.Second, func() (done bool, err error) {
+		return controller.hasPriorityLevelState(plcObj.Name), nil
+	})
+	if err != nil {
+		t.Errorf("expected the controller to reconcile the priority level configuration object: %s, error: %s", plcObj.Name, err)
+	}
+
+	reqInfo := &request.RequestInfo{
+		IsResourceRequest: false,
+		Path:              "/foobar",
+		Verb:              "GET",
+	}
+	noteFn := func(fs *flowcontrol.FlowSchema, plc *flowcontrol.PriorityLevelConfiguration, fd string) {}
+	workEstr := func() fcrequest.WorkEstimate { return fcrequest.WorkEstimate{InitialSeats: 1} }
+
+	flowUser := testUser{name: "test-user"}
+	rd := RequestDigest{
+		RequestInfo: reqInfo,
+		User:        flowUser,
+	}
+
+	// QueueNoteFn executes twice, once leaving the queue and once exiting the queue, which
+	// should result in ~500ms wait time in the latency tracker
+	qnf := fq.QueueNoteFn(func(bool) { time.Sleep(250 * time.Millisecond) })
+	ctx := request.WithLatencyTrackers(context.Background())
+	controller.Handle(ctx, rd, noteFn, workEstr, qnf, func() {})
+
+	latencyTracker, ok := request.LatencyTrackersFrom(ctx)
+	if !ok {
+		t.Fatalf("error getting latency tracker: %v", err)
+	}
+
+	// latency tracker should report ~500ms wait time (+5% margin of error)
+	latency := latencyTracker.APFQueueWaitTracker.GetLatency()
+	if latency.Milliseconds() < 500 || latency.Milliseconds() > 525 {
```

Comment:
I've seen that by setting 50ms the test started flaking for you.

I didn't look very deeply, but given we're computing the waiting times ourselves in APF filter and we're passing the clock to it, wouldn't using a `FakeClock` here (and changing qnf to instead to advance the clock by 250ms) make the test actually deterministic?

Reply:
Great suggestion -- I just tried this but I think it'll require a larger refactor because `FakeClock` does not implement the custom eventlock used by APF (`k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/eventclock`), but using the fake event clock (`k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/testing/eventclock`) is not compatible with `clock.Clock` required by `WithLatencyTrackersAndCustomClock`.

Any objections to fix this as a follow-up? It may require changing APF's eventclock implementation

Reply:
`durationTracker` is simply declared to be more demanding than it needs to be.  Just switch to

```
type durationTracker struct {
	clock             clock.PassiveClock
...
```

and there will be no incompatibility.

Reply:
Updating durationTracker to use a passive clock would work if we were calling `Track()`, but it doesn't since we're calling `TrackDuration()` which doesn't depend on `clock` and just receives the duration.

I was able to update the test to be determinstic, but it required replacing all references to `time.Now()` and `time.Since()` to use the passive clock in `configController`. PTAL